### PR TITLE
T23807

### PR DESCRIPTION
**Pull request by tchx84**
https://phabricator.endlessm.com/T23807

### CODE REVIEW

File: src/hackyballs/controlpanel.js
Comment (ptomato) [inline src/hackyballs/controlpanel.js]: I think it would be better to make the reset button a permanent part of the toolbox window, and have it emit a signal which is handled by your control panel.

File: src/hackyballs/controlpanel.js
Comment (ptomato) [inline src/hackyballs/controlpanel.js]: `this._model[prop.get_name()] = prop.default_value`?

File: src/hackyballs/level1.js
Comment (ptomato) [inline src/hackyballs/level1.js]: Since you are using `index` anyway to bind, you can give these any name you want. It might be better to be more descriptive.

File: src/hackyballs/level3.js
Comment (ptomato) [inline src/hackyballs/level3.js:95]: Is `pspec.default_value` the right value in all cases? In my case, a `Gdk.RGBA` (i.e. GBoxed) property's default value is always `null`, so I had to provide my own defaults...

File: src/hackyballs/model.js
Comment (ptomato) [inline src/hackyballs/model.js]: Is this meant to be addressed before merging?

**Review by ptomato — CHANGES_REQUESTED** [review]
I'm having trouble running it, I will comment on the Phabricator issue.

File: src/hackyballs/controlpanel.js
Comment (tchx84) [reply src/hackyballs/controlpanel.js]: what do you mean by toolbox window? You mean the fake one? Because the real one is not connected to the fake one.

File: src/hackyballs/controlpanel.js
Comment (tchx84) [reply src/hackyballs/controlpanel.js]: Remember the issue I mentioned in SF? Every time I use `this._model[name] = value`, the `notify` signal never gets emitted and everything gets out of sync ;/

File: src/hackyballs/level3.js
Comment (tchx84) [reply src/hackyballs/level3.js:95]: For hackyballs, yes. But is a valid point for whenever we try to generalize this code.

File: src/hackyballs/model.js
Comment (tchx84) [reply src/hackyballs/model.js]: It was my intention, but there is still one issue with how that function behaves, e.g. it will return '0' when the type passed is not strictly the same, like '10' and '10.1' it will still return '0' (and will print a CRITICAL). In other words, it only behaves properly, when the two values are strictly the same type.

So, I tried to complement it with `param_value_validate` but something weird happening with it as well.

```
const {GObject, Gtk} = imports.gi;

Gtk.init(null);

const w = new Gtk.Box();
const pspec = GObject.Object.find_property.call(w.constructor.$gtype, 'visible');
print(pspec);
print(pspec.owner_type);
print(pspec.default_value);
print(GObject.param_value_validate(pspec, pspec.default_value));
print(GObject.param_values_cmp(pspec, pspec.default_value, true));
print(GObject.param_values_cmp(pspec, pspec.default_value, false));
print(GObject.param_values_cmp(pspec, pspec.default_value, 'true'));
```

http://devdocs.baznga.org/gobject20~2.44p/gobject.param_values_cmp
http://devdocs.baznga.org/gobject20~2.44p/gobject.param_value_validate

So I would prefer not get blocked on this right now.

Comment (ptomato) [reply src/hackyballs/model.js]: I think I figured out that `param_values_cmp` and `param_value_validate` are currently unusable from GJS, because you don't have control over what type of GValue gets created, but to use the function correctly, the passed-in GValues must adhere to certain types. And there's no way to express that in g-i annotations.

I think it might be possible to make this function work correctly by overriding `GObject.ParamSpec.prototype.vfunc_values_cmp()` and just reimplementing the logic of the C function in JS, but I agree to not block on that. Would you mind opening a bug for GJS?

Comment (tchx84) [reply src/hackyballs/model.js]: Yes, ill report it.

Comment (tchx84) [reply src/hackyballs/model.js]: https://gitlab.gnome.org/GNOME/gjs/issues/210

File: src/framework/buttonGroup.js
Comment (ptomato) [inline src/framework/buttonGroup.js:7]: I would split this change out into a separate commit.

File: src/hackyballs/model.js
Comment (ptomato) [inline src/hackyballs/model.js]: Call this `ClippyIface` or something similar.

File: src/hackyballs/model.js
Comment (ptomato) [reply src/hackyballs/model.js]: I would remove the `XXX` then, to show that it's not something we really need to solve

File: src/hackyballs/model.js
Comment (ptomato) [inline src/hackyballs/model.js:37]: Why is it needed to split it into a base model and a derived model?

**Review by ptomato — CHANGES_REQUESTED** [review]
Just a few comments.

I pushed a small change to my branch as well, so it would be good to rebase on T23957 as well.

File: src/hackyballs/model.js
Comment (tchx84) [reply src/hackyballs/model.js:37]: I will need other models for different hack able things, and I need to be able to pass mapping data between app variables and my local variables.

File: src/hackyballs/sliders.js
Comment (ptomato) [inline src/hackyballs/sliders.js]: optional: The `_init` function can be left out if it only chains up.

File: src/hackyballs/controlpanel.js
Comment (ptomato) [inline src/hackyballs/controlpanel.js]: I made the same mistake in my original lockscreen PR and had to fix it later; when you bind the window you have to sync the initial lockscreens states with the window's unlock state, or else you will get a wrong state when you restart the app without restarting the hack toolbox service.

See commit 1b1c91ab7bc2ccded01b07cc4f66c0e838ac0d28

**Review by ptomato — CHANGES_REQUESTED** [review]
Only one important comment

**Review by ptomato — APPROVED** [review]

**Pull request merged**
